codepropertygraph/src/main/scala/io/shiftleft/passes/CpgPass.scala

```diff
@@ -168,6 +221,86 @@ abstract class NewStyleCpgPassBase[T <: AnyRef] extends CpgPassBase {
   }
 }
 
+abstract class NewStyleCpgPassBaseWithTimeout[T <: AnyRef](timeout: Long) extends CpgPassBase {
+  type DiffGraphBuilder = overflowdb.BatchedUpdate.DiffGraphBuilder
+
+  // generate Array of parts that can be processed in parallel
+  def generateParts(): Array[? <: AnyRef]
+
+  // setup large data structures, acquire external resources
+  def init(): Unit = {}
+
+  // release large data structures and external resources
+  def finish(): Unit = {}
+
+  // main function: add desired changes to builder
+  def runOnPart(builder: DiffGraphBuilder, part: T): Unit
+
+  // Override this to disable parallelism of passes. Useful for debugging.
+  def isParallel: Boolean = true
+
+  override def createAndApply(): Unit = createApplySerializeAndStore(null)
+
+  override def runWithBuilder(externalBuilder: BatchedUpdate.DiffGraphBuilder): Int = {
+    try {
+      init()
+      val parts  = generateParts()
+      val nParts = parts.size
+      nParts match {
+        case 0 =>
+        case 1 =>
+          runOnPart(externalBuilder, parts(0).asInstanceOf[T])
+        case _ =>
+          val stream =
+            if (!isParallel)
+              java.util.Arrays
+                .stream(parts)
+                .sequential()
+            else
+              java.util.Arrays
+                .stream(parts)
+                .parallel()
+
+          implicit val ec: scala.concurrent.ExecutionContext = scala.concurrent.ExecutionContext.global
+          val stopAt                                         = System.currentTimeMillis() + timeout * 1000
+          var exitByTimeout                                  = false
+          val diffGraphAccumulator                           = Cpg.newDiffGraphBuilder
+          breakable {
+            stream.forEach { part =>
+              val currentTimeInMs = System.currentTimeMillis()
+              if (timeout == -1 || currentTimeInMs < stopAt) {
+                val future = Future {
+                  val diffGraphBuilder = Cpg.newDiffGraphBuilder
+                  runOnPart(diffGraphBuilder, part.asInstanceOf[T])
+                  diffGraphBuilder
+                }
+                val duration =
+                  if timeout == -1 then Duration.Inf else Duration(stopAt - currentTimeInMs, TimeUnit.MILLISECONDS)
+                Try(Await.result(future, duration)) match {
```

Comment:
You have taken care of handling the overall timeout by honouring the remaining timeout at the thread level as well. 

We need to use this place to introduce the thread-level timeout.

Reply:
This is really not parallelising the processing. 

`stream.forEach` creates one `Future` (at line no 272) for a given `part` and then the current main thread will wait for that future to finish (at line no 279). So In principle, you are processing the `part` sequentially. Which is as good as not using the `Future`

Reply:
In order to implement the parallelism you need to iterate over all parts and start all the `Futures` and collect them in collection and apply timeout on Seqeunce of Futures.

Refer this code snippet as sample - https://github.com/Privado-Inc/privado-core/blob/main/src/main/scala/ai/privado/utility/ConcurrentProcessor.scala#L94-L103